### PR TITLE
Fix #1299: FIDO2: Check AAGUID during registration

### LINKED ISSUE #1299: FIDO2: Check AAGUID during registration
Add an option to configure allowed AAGUIDs which can be used when registering FIDO2 tokens. This configuration should be bound to the application.

Additionally, allow to configure accepted attestation formats and check them during registration.

### PR DESCRIPTION
**Pull request by romanstrobl**
Missing tasks:
- [x] tests for AAGUID check and attestation format check
- [x] validate known configuration keys
- [x] update documentation

### CODE REVIEW

File: powerauth-client-model/src/main/java/com/wultra/security/powerauth/client/PowerAuthClient.java [hunk 1]
```diff
@@ -2135,4 +2135,79 @@ RecoveryCodeActivationResponse createActivationUsingRecoveryCode(String recovery
...
+
+    /**
+     * Create an application configuration.
+     * @param applicationId Application identifier.
+     * @param key Configuration key.
+     * @param values Configuration values.
+     * @return Create application configuration response.
+     * @throws PowerAuthClientException In case REST API call fails.
+     */
+    CreateApplicationConfigResponse createApplicationConfig(String applicationId, String key, List<String> values) throws PowerAuthClientException;
```
Comment (banterCZ) [inline powerauth-client-model/src/main/java/com/wultra/security/powerauth/client/PowerAuthClient.java:2162]: What about keeping only the method with `CreateApplicationConfigRequest` and not listing all arguments?

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/controller/api/PowerAuthController.java [hunk 2]
```diff
@@ -433,6 +433,43 @@ public ObjectResponse<SupportApplicationVersionResponse> supportApplicationVersi
...
+    /**
+     * Call {@link PowerAuthService#getApplicationConfig(GetApplicationConfigRequest)} method and
+     * return the response.
+     * @param request Get application configuration.
+     * @return Application configuration response.
+     * @throws Exception In case the service throws exception.
+     */
+    @PostMapping("/application/config/detail")
+    public ObjectResponse<GetApplicationConfigResponse> getApplicationConfig(@RequestBody ObjectRequest<GetApplicationConfigRequest> request) throws Exception {
+        return new ObjectResponse<>("OK", powerAuthService.getApplicationConfig(request.getRequestObject()));
```
Comment (banterCZ) [inline powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/controller/api/PowerAuthController.java:445]: The constructor with object value only adds implicitly the status `OK`

```suggestion
        return new ObjectResponse<>(powerAuthService.getApplicationConfig(request.getRequestObject()));
```

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/database/model/converter/ListToJsonConverter.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+    @Override
+    public String convertToDatabaseColumn(List<String> attributes) {
+        if (attributes == null) {
+            return EMPTY_LIST;
+        }
+        try {
+            return objectMapper.writeValueAsString(attributes);
+        } catch (JsonProcessingException ex) {
+            logger.warn("Conversion failed for attribute list, error: " + ex.getMessage(), ex);
```
Comment (banterCZ) [inline powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/database/model/converter/ListToJsonConverter.java:64]:
```suggestion
            logger.warn("Conversion failed for attribute list, error: {}", ex.getMessage(), ex);
```

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/database/model/entity/ApplicationConfigEntity.java [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    public Long getRid() {
+        return rid;
+    }
+
+    /**
+     * Set application configuration RID.
+     *
+     * @param id Application configuration RID.
+     */
+    public void setRid(Long id) {
```
Comment (banterCZ) [inline powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/database/model/entity/ApplicationConfigEntity.java:93]: May we use Lombok getter/setter?

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/ApplicationConfigServiceBehavior.java [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * Behavior class implementing management of application configuration.
+ *
+ * @author Roman Strobl, roman.strobl@wultra.com
+ */
+@Component("applicationConfigServiceBehavior")
```
Comment (banterCZ) [inline powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/ApplicationConfigServiceBehavior.java:48]: Is the value really needed? This name should be the same by default. Even though it is initialized to `new ArrayList<>()`.

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/ApplicationConfigServiceBehavior.java [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+            final ApplicationConfigEntity config = new ApplicationConfigEntity();
+            config.setApplication(appEntity);
+            config.setKey(key);
+            config.setValues(values);
+            configRepository.save(config);
+        });
+        final CreateApplicationConfigResponse response = new CreateApplicationConfigResponse();
+        response.setApplicationId(applicationId);
+        response.setKey(key);
+        response.setValues(List.copyOf(values));
```
Comment (banterCZ) [inline powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/ApplicationConfigServiceBehavior.java:140]: 🤔 Is it really NPE safe? I afraid that it is possible to send explicit `null` in the JSON.

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/fido2/PowerAuthRegistrationProvider.java [hunk 7]
```diff
@@ -141,4 +149,39 @@ public void revokeRegistrationByChallengeValue(String applicationId, String chal
...
+        final GetApplicationConfigResponse configResponse = configService.getApplicationConfig(configRequest);
+        final String aaguidStr = new String(aaguid, StandardCharsets.UTF_8);
+        Optional<ApplicationConfigurationItem> configFmt = configResponse.getApplicationConfigs().stream()
+                .filter(cfg -> CONFIG_KEY_ALLOWED_ATTESTATION_FMT.equals(cfg.getKey()))
+                .findFirst();
+
+        if (configFmt.isPresent()) {
+            List<String> allowedFmts = configFmt.get().getValues();
+            if (!allowedFmts.contains(attestationFormat)) {
+                logger.warn("Rejected attestation format for FIDO2 registration: {}" + attestationFormat);
```
Comment (banterCZ) [inline powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/fido2/PowerAuthRegistrationProvider.java:168]:
```suggestion
                logger.warn("Rejected attestation format for FIDO2 registration: {}", attestationFormat);
```

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/fido2/PowerAuthRegistrationProvider.java [hunk 8]
```diff
@@ -141,4 +149,39 @@ public void revokeRegistrationByChallengeValue(String applicationId, String chal
...
+        }
+
+        Optional<ApplicationConfigurationItem> configAaguids = configResponse.getApplicationConfigs().stream()
+                .filter(cfg -> CONFIG_KEY_ALLOWED_AAGUIDS.equals(cfg.getKey()))
+                .findFirst();
+
+        if (configAaguids.isPresent()) {
+            List<String> allowedAaguids = configAaguids.get().getValues();
+            if (!allowedAaguids.contains(aaguidStr)) {
+                logger.warn("Rejected AAGUID value for FIDO2 registration: {}" + aaguidStr);
```
Comment (banterCZ) [inline powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/fido2/PowerAuthRegistrationProvider.java:180]:
```suggestion
                logger.warn("Rejected AAGUID value for FIDO2 registration: {}", aaguidStr);
```

File: powerauth-client-model/src/main/java/com/wultra/security/powerauth/client/PowerAuthClient.java (hunk 1 above)
Comment (romanstrobl) [reply powerauth-client-model/src/main/java/com/wultra/security/powerauth/client/PowerAuthClient.java:2162]: I did this for consistency with other methods in the client, if there are few arguments, we provide such methods.

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/ApplicationConfigServiceBehavior.java (hunk 5 above)
Comment (romanstrobl) [reply powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/ApplicationConfigServiceBehavior.java:48]: We can remove the qualifier, sure.

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/ApplicationConfigServiceBehavior.java (hunk 6 above)
Comment (romanstrobl) [reply powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/ApplicationConfigServiceBehavior.java:140]: I can add the null check.

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/controller/api/PowerAuthController.java [hunk 9]
```diff
@@ -433,6 +433,43 @@ public ObjectResponse<SupportApplicationVersionResponse> supportApplicationVersi
...
+    /**
+     * Call {@link PowerAuthService#createApplicationConfig(CreateApplicationConfigRequest)} method and
+     * return the response.
+     * @param request Create an application configuration.
+     * @return Create application configuration response.
+     * @throws Exception In case the service throws exception.
+     */
+    @PostMapping("/application/config/create")
+    public ObjectResponse<CreateApplicationConfigResponse> createApplicationConfig(@RequestBody ObjectRequest<CreateApplicationConfigRequest> request) throws Exception {
+        return new ObjectResponse<>("OK", powerAuthService.createApplicationConfig(request.getRequestObject()));
```
Comment (banterCZ) [inline powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/controller/api/PowerAuthController.java:457]:
```suggestion
        return new ObjectResponse<>(powerAuthService.createApplicationConfig(request.getRequestObject()));
```

File: powerauth-fido2/src/main/java/com/wultra/powerauth/fido2/rest/model/enumeration/Fido2ConfigKeys.java [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package com.wultra.powerauth.fido2.rest.model.enumeration;
+
+/**
+ * FIDO2 configuration key strings.
```
Comment (banterCZ) [inline powerauth-fido2/src/main/java/com/wultra/powerauth/fido2/rest/model/enumeration/Fido2ConfigKeys.java:23]: Missing author tag.

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/database/model/converter/ListToJsonConverter.java [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+
+    @Override
+    public List<String> convertToEntityAttribute(String dbValue) {
+        if (dbValue == null) {
+            return null;
+        }
+        try {
+            return objectMapper.readValue(dbValue, new TypeReference<>() {});
+        } catch (JsonProcessingException ex) {
+            logger.warn("Conversion failed for attribute list, error: " + ex.getMessage(), ex);
```
Comment (banterCZ) [inline powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/database/model/converter/ListToJsonConverter.java:77]:
```suggestion
            logger.warn("Conversion failed for attribute list, error: {}", ex.getMessage(), ex);
```

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/database/model/entity/ApplicationConfigEntity.java [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+
+    /**
+     * Constructor for a new application configuration.
+     *
+     * @param application Application entity.
+     * @param key         Configuration key.
+     * @param values       Configuration values.
+     */
+    public ApplicationConfigEntity(ApplicationEntity application, String key, List<String> values) {
+        super();
```
Comment (banterCZ) [inline powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/database/model/entity/ApplicationConfigEntity.java:77]: `supper()` is called implicitly.
```suggestion
```

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/ApplicationConfigServiceBehavior.java [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+
+/**
+ * Behavior class implementing management of application configuration.
+ *
+ * @author Roman Strobl, roman.strobl@wultra.com
+ */
+@Component
+public class ApplicationConfigServiceBehavior {
+
+    private static final Logger logger = LoggerFactory.getLogger(ApplicationConfigServiceBehavior.class);
```
Comment (banterCZ) [inline powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/ApplicationConfigServiceBehavior.java:54]: You may use `@Slf4j` if you want to.

**Review by banterCZ — DISMISSED** [review]
Thanks for improving the documentation.

**Review by banterCZ — APPROVED** [review]
Thanks for reflecting my comments.

**Pull request merged**